### PR TITLE
Refactor/layout

### PR DESCRIPTION
**Pull request by venjiang**
refactor project layout

---
This change is [Reviewable](https://reviewable.io/reviews/yomorun/yomo/235)

### CODE REVIEW

File: source.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	SourceLogPrefix = "\033[32m[yomo:source]\033[0m "
+)
+
+type Source interface {
+	Write(p []byte) (n int, err error)
+	SetDataTag(tag uint8)
+	Close() error
+	Connect() error
+	WriteWithTag(tag uint8, data []byte) error
+	WriteWithTransaction(transactionID string, tag uint8, data []byte) error
```
Comment (xiaojian-hong) [inline source.go:23]: 感觉 `WriteWithTag` 和  `WriteWithTransaction` 可以合并到 `Write` 方法里，通过`options` 模式来灵活配置参数？
比如 `Write(data []byte, WithTag(0x10), WithTransactionID("tid"))`

File: internal/core/server.go [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+	// counter +1
+	atomic.AddInt64(&s.counterOfDataFrame, 1)
+	// 收到数据帧
+	currentSfn := f.Issuer()
+	logger.Infof("%sframeType=%s, issuer=%s, counter=%d", ServerLogPrefix, f.Type(), currentSfn, s.counterOfDataFrame)
+	// 因为是Immutable Stream，按照规则发送给 sfn
+	var j int
+	for i, k := range s.funcBuckets {
+		// 发送给 currentSfn 的下一个 sfn
+		if k == currentSfn {
```
[the remark this answers is not in the document]
Comment (venjiang) [reply internal/core/server.go:199]: 好像可以哦

File: source.go (hunk 1 above)
Comment (venjiang) [reply source.go:23]: 是的，有一些类似可以用Options模式解决的，后面还是有很多需要优化的地方

File: severless.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
+package yomo
```
Comment (fanweixiao) [inline severless.go:1]: Shall we rename this file to `sfn.go`?

File: severless.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+	"fmt"
+
+	"github.com/yomorun/yomo/internal/core"
+	"github.com/yomorun/yomo/internal/frame"
+	"github.com/yomorun/yomo/logger"
+	"github.com/yomorun/yomo/zipper/tracing"
+)
+
+const (
+	streamFunctionLogPrefix = "\033[31m[yomo:sfn]\033[0m "
```
Comment (fanweixiao) [inline severless.go:14]: changed from `StreamFunctionLogPrefix` to `streamFunctionLogPrefix`, hide to end user

File: severless.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+	// SetHandler set the handler function, which accept the raw bytes data and return the tag & response
+	SetHandler(fn func([]byte) (byte, []byte)) error
+	// Connect create a connection to the zipper
+	Connect() error
+	// Close will close the connection
+	Close() error
+}
+
+// NewStreamFunction create a stream function.
+func NewStreamFunction(name string, opts ...Option) StreamFunction {
```
Comment (fanweixiao) [inline severless.go:30]: return `StreamFunction` interface instead of `streamFunction` struct.

File: severless.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+func (s *streamFunction) SetHandler(fn func([]byte) (byte, []byte)) error {
+	s.fn = fn
+	logger.Debugf("%sSetHandler(%v)", streamFunctionLogPrefix, s.fn)
+	return nil
+}
+
+// Connect create a connection to the zipper, when data arrvied, the data will be passed to the
+// handler which setted by SetHandler method.
+func (s *streamFunction) Connect() error {
+	// TODO: should guarante SetHandler() has been called
```
Comment (fanweixiao) [inline severless.go:70]: shall we check if `SetHandler()` is executed before this func?

**Review by fanweixiao — CHANGES_REQUESTED** [review]
Review `severless.go`

File: severless.go (hunk 4 above)
Comment (venjiang) [reply severless.go:14]: This is temporary, and the Logger needs to be refactored later.

File: severless.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+	logger.Debugf("%s Connect()", StreamFunctionLogPrefix)
+	// notify underlying network operations, when data with tag we observed arrived, invoke the func
+	s.client.SetDataFrameObserver(func(tag byte, carraige []byte, metaFrame MetaFrame) {
+		for _, t := range s.observed {
+			if t == tag {
+				logger.Debugf("%sreceive DataFrame, tag=%# x, carraige=%# x", StreamFunctionLogPrefix, tag, carraige)
+				s.onDataFrame(carraige, metaFrame)
+				// TODO: consider remove return
+				return
```
[the remark this answers is not in the document]
Comment (xiaojian-hong) [reply severless.go:79]: Yes, as we support multiple data tags, we can remove this `return`.
Another question, all data tags will use the same `Handler`? Or each data tag will use an individual `Handler`?

File: severless.go (hunk 6 above)
Comment (fanweixiao) [reply severless.go:70]: `cli` will take charge of this.

File: zipper.go [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+	server            *core.Server
+	client            *core.Client
+	downstreamZippers []Zipper
+	// logger            utils.Logger
+}
+
+var _ Zipper = &zipper{}
+
+// NewZipperWithOptions create a zipper instance.
+func NewZipperWithOptions(name string, opts ...Option) Zipper {
```
Comment (venjiang) [inline zipper.go:66]: NewZipperWithOptions and NewZipper function can be combined into one.

File: zipper.go [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+	options := newOptions(opts...)
+	return createZipperServer(name, options.ZipperAddr)
+}
+
+// NewZipper create a zipper instance from config files.
+func NewZipper(conf string) (Zipper, error) {
```
Comment (venjiang) [inline zipper.go:72]: config yaml file can use `WithConfig` options

File: internal/util/zipper_workflow.go [hunk 10]
```diff
@@ -0,0 +1,119 @@
+package util
```
Comment (venjiang) [inline internal/util/zipper_workflow.go:1]: it be better to use `config` as the package name?

**Review by fanweixiao — APPROVED** [review]

**Pull request merged**